### ISSUE TITLE
Career mapping update (5)

### ISSUE THREAD

**Issue by nihonjinrxs** [loyno-mathcs/workforwardnola#5]
As a user seeking assistance I want to be able to be matched with careers I can get assistance with so that I can find a job.

**Comment by cam-carter** [comment]
BLOCKER:  Need career descriptions from client

**Comment by nihonjinrxs** [comment]
@tungly @bsabo Note the blocker on this issue. Some information is missing from the spreadsheet to be able to do this.  The content in the `description` and `foundational skills` columns in the [Google Sheet here](https://docs.google.com/spreadsheets/d/1lMeZZ_vu-xIVKLgo7obKlF0X4iJ3oHkLKl8uIrPsSt8/pubhtml) are not available for the new careers in the [spreadsheet provided](https://www.dropbox.com/personal/2107F-COSCA451-ProjectShare?preview=CFA+questions+(3).xlsx).

@cam-carter Note the difference in my report of the problem to the clients and yours -- you need to be more specific and tag them to ensure they see the message.  For now, please use [the Lorem ipsum generator](http://www.lipsum.com/) to generate dummy text so that you can complete the task with a placeholder, and you can swap in the actual content once you get it.

**Comment by nihonjinrxs** [comment]
@cam-carter New spreadsheet has been uploaded to the Dropbox: `CFA questions 922 (1).xlsx`.

**Comment by tmloupe** [comment]
@nihonjinrxs @tungly @bsabo There's a permissions issue with the spreadsheet link.  The site uses spreadsheet links to update the content. We are unable to use any other spreadsheet besides the one provided, and our Loyno gmail accounts will not work.

**Comment by nihonjinrxs** [comment]
@tmloupe I've requested access to the spreadsheet from the current owners, but have not received it yet.  Note that you can use a different spreadsheet link if you create it following these instructions in the docs, so this should not be a blocker: https://github.com/loyno-mathcs/workforwardnola/blob/master/docs/career_assessment_how_to.md#starting-from-scratch.

**Comment by nihonjinrxs** [comment]
@tmloupe Also, it might be worthwhile to trace the spreadsheet upload code, so that you better understand how that data gets to the app database.

- The route `/manage` is [defined here](https://github.com/loyno-mathcs/workforwardnola/blob/master/app.rb#L141-L145).
- The upload form rendered when you go to `/manage` is created via [the template here](https://github.com/loyno-mathcs/workforwardnola/blob/master/templates/manage.mustache).
- The template uses [the JavaScript here](https://github.com/loyno-mathcs/workforwardnola/blob/master/assets/javascript/spreadsheets.js) in the client browser to handle validation and upload.
- Career data is updated in the database in [the Career model here](https://github.com/loyno-mathcs/workforwardnola/blob/master/models/career.rb#L6-L28).
- Trait data is updated in the database in [the Trait model here](https://github.com/loyno-mathcs/workforwardnola/blob/master/models/trait.rb#L5-L12).

**Comment by tmloupe** [comment]
@nihonjinrxs I have tried going into manage. I have been getting errors trying to use my own spreadsheets on a non-Loyno account (which is all I can use for the manage page because of the Loyno restriction). I get a Connection error, which can be seen in the JavaScript file, but there is no documentation for it; it simply says to try again. Dummy data does not get accepted by the link. I will have to investigate again, but it does not seem to accept my spreadsheet. The regex may need to be changed, too, as the spreadsheet link google sheets is giving me is different than the one that is given by the regex. I neglected to mention this in my previous comment; I apologize. I will try to solve this ASAP, as we cannot do anything if we cannot link even the template - the salary wages will additionally not work until we get a template running that accommodates salary wages.

**Comment by nihonjinrxs** [comment]
Per planning session, needs the following subtasks:

- [x] Figure out how to get new data into the system
- [ ] Get new assessment questions and associations to new traits
- [ ] Update spreadsheet with new data
- [ ] Modify code for assessment questions

**Comment by bsabo** [comment]
Just uploaded all the final photos and new spreadsheet with a total of 11 questions that follow the “me/not me” format. I must apologize for the photo ID’s. That’s how they come from the stock photo service so people can do searches, but I think were written by Donald Trump.

**Comment by bsabo** [comment]
Can you do me a favor and give me the names of the class. I’d like to be able to refer to them by name.

Skinny guy beard
Heavier guy beard
Kid with baseball cap
Kid who forgets his password
Girl.

Thanks.

**Comment by tmloupe** [comment]
@bsabo Concerning the logic of the assessment questions: Would you like to map certain questions/traits to link to a "reverse" trait? For example, if one chooses "not me" for teamwork, would you like to map it to a different trait? This would help find a more accurate career, as the code checks the first three traits chosen - by choosing reverse traits, it will allow for higher scores on certain traits for careers. If so, would you be willing to link questions to have "reverse" traits? Note that not each question needs to have a reverse. If not, then the code might be a little weird, as we would need to create logic for not having any traits at all, a.k.a the worst case scenario.

**Comment by bsabo** [comment]
I am working with Tung on understanding the concept better. Should have an snwer soon.

**Comment by tmloupe** [comment]
@bsabo @nihonjinrxs @tungly  Hi, I apologize for asking again, but do you have a conclusion on whether or not you would like the assessment changed to include "not me" traits based on the questions? After coding, it is becoming increasingly apparent that this may pose an issue to the accuracy of the career mapping. To clarify from our point of view:

1. The user encounters a question, with a "me" or "not me"
2. The user selects "me" or "not me", assigning a point to a trait score (+1 for the trait)
3. The logic adds the traits up and matches the user with the careers with the most traits that are similar.
If a career has multiple traits (for example, if a career has 3 traits) and a user has all 3 traits, then the code will consider that to be a better match than one that has 2 of the same traits (though they  may both still appear).

What I would like to know is: do you want to follow this same logic, which would include mapping multple questions to careers, or would you prefer the assessment to be 1 trait per question? As such, "not me" would be useful for assigning a trait that gets mapped on to a career - it also eliminates the possibility of having so few traits that the careers are inaccurate, or so many that it is also inaccurate (it will only display the first 3). Additionally, when using no reverse traits, all "not me" responses will break the website, which, ideally we would like to not happen. If you would like further clarification, we could talk about it next sprint and we can keep what we have for now (it's still different than the original).

**Comment by bsabo** [comment]
Yes, let’s keep that. I met with Tung yesterday and will send you reverse traits today

**Comment by bsabo** [comment]
I've shared an item with you:

CFA questions 923a WITH PHOTO new reverse traits
https://docs.google.com/spreadsheets/d/1-RytFdPlVaREcLll4neJ_60inUwuT6MPdu3Hh1aY7qg/edit?usp=sharing&ts=59e0c9cb

It's not an attachment -- it's stored online. To open this item, just click  
the link above.

I am sending the google doc link. The download to Excel looks strange so I  
wanted to make sure you got the who document. I can resend from Excel if  
you like.

**Comment by maahern** [comment]
@tmloupe This comment is in continuation of the conversation from the GroupMe chat. The method error was due to the reference of to_money(). This method is no longer used, as we changed experienced_range to a string from an integer. I made the update and commit the changes. The careers page is still generating an error, however. Which google doc are you using? The issue may be that experienced_wage (integer) is now experienced_range (string). Let me know. I am having some errors on my end regarding postgres importing entry_wage data. I'll save the functioning changes and push them so that everyone has the proper schema to use the primary database.
